File: sol/solution/0001-0100/0016/Solution.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
    nums.sort!
    n = nums.length
    closest_sum = nums[0] + nums[1] + nums[2]
    (0...(n-2)).each do |i|
        left = i + 1
        right = n - 1
        while left < right
            curr_sum = nums[i] + nums[left] + nums[right]
            return curr_sum if curr_sum == target
            closest_sum = curr_sum if (curr_sum - target).abs < (closest_sum - target).abs
            if curr_sum < target
                left += 1
            else
                right -= 1
            end
        end
    end
    closest_sum
end

File: sol/solution/0001-0100/0016/Solution.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size(), closestSum = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < n - 2; i++) {
            int left = i + 1, right = n - 1;
            while (left < right) {
                int currSum = nums[i] + nums[left] + nums[right];
                if (currSum == target) return currSum;
                if (abs(currSum - target) < abs(closestSum - target)) {
                    closestSum = currSum;
                }
                if (currSum < target) left++;
                else right--;
            }
        }
        return closestSum;
    }
};